File: aura-your-ai-executive-assistant-main/src/lib/googleCalendar.ts
import { getGoogleAccessToken } from './googleAuth';

const CALENDAR_API_BASE = 'https://www.googleapis.com/calendar/v3';

export interface CalendarEvent {
  id?: string;
  summary: string;
  description?: string;
  start: {
    dateTime: string;
    timeZone?: string;
  };
  end: {
    dateTime: string;
    timeZone?: string;
  };
  attendees?: Array<{ email: string }>;
  location?: string;
  reminders?: {
    useDefault: boolean;
    overrides?: Array<{
      method: 'email' | 'popup';
      minutes: number;
    }>;
  };
}

export async function createCalendarEvent(event: CalendarEvent): Promise<any> {
  const accessToken = getGoogleAccessToken();
  
  if (!accessToken) {
    throw new Error('Not authenticated with Google');
  }

  try {
    const response = await fetch(
      `${CALENDAR_API_BASE}/calendars/primary/events`,
      {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          ...event,
          reminders: event.reminders || {
            useDefault: false,
            overrides: [
              { method: 'email', minutes: 60 },
              { method: 'popup', minutes: 30 },
              { method: 'popup', minutes: 10 },
            ],
          },
        }),
      }
    );

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error?.message || 'Failed to create calendar event');
    }

    return await response.json();
  } catch (error) {
    console.error('Error creating calendar event:', error);
    throw error;
  }
}

export async function updateCalendarEvent(
  eventId: string,
  updates: Partial<CalendarEvent>
): Promise<any> {
  const accessToken = getGoogleAccessToken();
  
  if (!accessToken) {
    throw new Error('Not authenticated with Google');
  }

  try {
    const response = await fetch(
      `${CALENDAR_API_BASE}/calendars/primary/events/${eventId}`,
      {
        method: 'PATCH',
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(updates),
      }
    );

    if (!response.ok) {
      throw new Error('Failed to update calendar event');
    }

    return await response.json();
  } catch (error) {
    console.error('Error updating calendar event:', error);
    throw error;
  }
}

export async function deleteCalendarEvent(eventId: string): Promise<void> {
  const accessToken = getGoogleAccessToken();
  
  if (!accessToken) {
    throw new Error('Not authenticated with Google');
  }

  try {
    const response = await fetch(
      `${CALENDAR_API_BASE}/calendars/primary/events/${eventId}`,
      {
        method: 'DELETE',
        headers: {
          'Authorization': `Bearer ${accessToken}`,
        },
      }
    );

    if (!response.ok) {
      throw new Error('Failed to delete calendar event');
    }
  } catch (error) {
    console.error('Error deleting calendar event:', error);
    throw error;
  }
}

export async function listCalendarEvents(
  timeMin?: string,
  timeMax?: string
): Promise<any[]> {
  const accessToken = getGoogleAccessToken();
  
  if (!accessToken) {
    throw new Error('Not authenticated with Google');
  }

  try {
    const params = new URLSearchParams({
      singleEvents: 'true',
      orderBy: 'startTime',
    });

    if (timeMin) params.append('timeMin', timeMin);
    if (timeMax) params.append('timeMax', timeMax);

    const response = await fetch(
      `${CALENDAR_API_BASE}/calendars/primary/events?${params}`,
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
        },
      }
    );

    if (!response.ok) {
      throw new Error('Failed to fetch calendar events');
    }

    const data = await response.json();
    return data.items || [];
  } catch (error) {
    console.error('Error fetching calendar events:', error);
    throw error;
  }
}

File: backend/services/google_calendar.py
from .google_auth import google_auth
from datetime import datetime, timedelta

class GoogleCalendarService:
    def __init__(self):
        self.calendar_id = 'primary'

    def get_service(self):
        return google_auth.get_service('calendar', 'v3')

    def check_conflicts(self, start_time_iso, duration_minutes=30):
        """Checks if there's any event overlapping with the requested time."""
        service = self.get_service()
        if not service:
            return True # Assume conflict if service is unavailable

        start_dt = datetime.fromisoformat(start_time_iso)
        end_dt = start_dt + timedelta(minutes=duration_minutes)

        time_min = start_dt.isoformat() + 'Z'
        time_max = end_dt.isoformat() + 'Z'

        try:
            events_result = service.events().list(
                calendarId=self.calendar_id,
                timeMin=time_min,
                timeMax=time_max,
                singleEvents=True
            ).execute()
            events = events_result.get('items', [])
            return len(events) > 0
        except Exception as e:
            print(f"Error checking conflicts: {e}")
            return True

    def create_event(self, summary, description, start_time_iso, duration_minutes=30, location=None):
        """Creates a Google Calendar event."""
        service = self.get_service()
        if not service:
            return None

        start_dt = datetime.fromisoformat(start_time_iso)
        end_dt = start_dt + timedelta(minutes=duration_minutes)

        event = {
            'summary': summary,
            'description': description,
            'start': {
                'dateTime': start_dt.isoformat(),
                'timeZone': 'UTC',
            },
            'end': {
                'dateTime': end_dt.isoformat(),
                'timeZone': 'UTC',
            },
            'location': location,
            'reminders': {
                'useDefault': False,
                'overrides': [
                    {'method': 'email', 'minutes': 24 * 60},
                    {'method': 'popup', 'minutes': 10},
                ],
            },
        }

        try:
            event = service.events().insert(calendarId=self.calendar_id, body=event).execute()
            return event.get('id')
        except Exception as e:
            print(f"Error creating event: {e}")
            return None

google_calendar = GoogleCalendarService()
